src/components/tl/ucp/reduce_scatter/reduce_scatter_ring.c

```diff
@@ -0,0 +1,362 @@
+    ucc_ring_frag_block_offset(task, count, send_block, &block_offset,
+                               &frag_offset);
+    UCPCHECK_GOTO(ucc_tl_ucp_send_nb(
+                      PTR_OFFSET(sbuf, (block_offset + frag_offset) * dt_size),
+                      frag_count * dt_size, mem_type, sendto, team, task),
+                  task, out);
+
+    status = ucc_tl_ucp_reduce_scatter_ring_progress(&task->super);
+    if (UCC_INPROGRESS == status) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+    return ucc_task_complete(coll_task);
+out:
+    return task->super.super.status;
+}
+
+static ucc_status_t
+ucc_tl_ucp_reduce_scatter_ring_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+
+    if (task->reduce_scatter_ring.inv_map.type != UCC_EP_MAP_FULL) {
+        ucc_ep_map_destroy(&task->reduce_scatter_ring.inv_map);
+    }
+    ucc_mc_free(task->reduce_scatter_ring.scratch_mc_header);
+    return ucc_tl_ucp_coll_finalize(coll_task);
+}
+
+static ucc_status_t ucc_tl_ucp_reduce_scatter_ring_init_subset(
+    ucc_base_coll_args_t *coll_args, ucc_base_team_t *team,
+    ucc_coll_task_t **task_h, ucc_subset_t subset, int n_frags, int frag)
+{
+    ucc_tl_ucp_team_t *tl_team  = ucc_derived_of(team, ucc_tl_ucp_team_t);
+    ucc_rank_t         size     = UCC_TL_TEAM_SIZE(tl_team);
+    size_t             count    = coll_args->args.dst.info.count * size;
+    ucc_datatype_t     dt       = coll_args->args.dst.info.datatype;
+    size_t             dt_size  = ucc_dt_size(dt);
+    ucc_memory_type_t  mem_type = coll_args->args.dst.info.mem_type;
+    size_t             to_alloc, max_segcount;
+    ucc_tl_ucp_task_t *task;
+    ucc_status_t       status;
+
+    task                 = ucc_tl_ucp_init_task(coll_args, team);
+    task->super.post     = ucc_tl_ucp_reduce_scatter_ring_start;
+    task->super.progress = ucc_tl_ucp_reduce_scatter_ring_progress;
+    task->super.finalize = ucc_tl_ucp_reduce_scatter_ring_finalize;
+    task->subset         = subset;
+
+    if (task->subset.map.type != UCC_EP_MAP_FULL) {
+        status = ucc_ep_map_create_inverse(task->subset.map,
+                                           &task->reduce_scatter_ring.inv_map);
+        if (UCC_OK != status) {
+            return status;
+        }
+    } else {
+        task->reduce_scatter_ring.inv_map.type   = UCC_EP_MAP_FULL;
+        task->reduce_scatter_ring.inv_map.ep_num = task->subset.map.ep_num;
+    }
+    task->reduce_scatter_ring.n_frags = n_frags;
+    task->reduce_scatter_ring.frag    = frag;
+
+    if (UCC_IS_INPLACE(coll_args->args)) {
+        count /= size;
+    }
+
+    max_segcount = ucc_buffer_block_count(count, size, 0);
+    //TODO need less mem - dvivide by n_frags
+    to_alloc = max_segcount + (size > 2 ? 2 * max_segcount : 0);
+    status   = ucc_mc_alloc(&task->reduce_scatter_ring.scratch_mc_header,
```

Comment:
I think it's better to allocate one chunk of memory and give portion of it to each subset rather than allow each subset allocate it's own scratch

Reply:
agree. done.